### PR TITLE
Refactor code to use new binary_serializer

### PR DESCRIPTION
**Pull request by jakobod**
This PR refactors the incubator code base to use the new `binary_serializer`.

### CODE REVIEW

**Comment by jakobod** [comment]
Two things:
1. I had to include `message.hpp` and `make_message.hpp` in two places because `make_error` would not  compile without those headers. This seems like somewhat of a hacky solution.. 
2. The `caf::span` type is not implicitly casted from `span<T>` to `span<const T>` anymore. Hence I had to add lots of `as_bytes(make_span(x))` calls. This seems unwieldy..

**Comment by Neverlord** [comment]
> This seems like somewhat of a hacky solution..

I think working with forward declarations whenever possible is generally a good thing. Sure, it adds some "inconvenience" to the call side by having to include a couple extra headers, but I think it's worth it.

> his seems unwieldy..

Agree. Please fix this upstream rather than working around it multiple times. 🙂

File: libcaf_net/caf/net/stream_transport.hpp [hunk 1]
```diff
@@ -84,7 +84,9 @@ class stream_transport : public stream_transport_base<Application> {
                     << CAF_ARG(this->handle_.id) << CAF_ARG(*num_bytes));
       this->collected_ += *num_bytes;
       if (this->collected_ >= this->read_threshold_) {
-        if (auto err = this->next_layer_.handle_data(*this, this->read_buf_)) {
+        if (auto err = this->next_layer_.handle_data(*this,
+                                                     make_span(
```
Comment (Neverlord) [inline libcaf_net/caf/net/stream_transport.hpp]: You've removed the `as_bytes`, but this should also compile without `make_span`, does it not?

Comment (jakobod) [reply libcaf_net/caf/net/stream_transport.hpp]: It should compile without `make_span` but isn't it clearer like this?

Comment (Neverlord) [reply libcaf_net/caf/net/stream_transport.hpp]: I don't think so. 😉

The function takes a span, which is just a view into *some* sort of buffer. Just like we wouldn't manually convert a string to a `string_view`, this only adds extra noise and verbosity.

Comment (jakobod) [reply libcaf_net/caf/net/stream_transport.hpp]: Alright, I get that. I will remove those 👍

Comment (jakobod) [reply libcaf_net/caf/net/stream_transport.hpp]: Done.

**Pull request merged**